### PR TITLE
Refactor all machine

### PR DESCRIPTION
**Pull request by LiuYuHui**

### CODE REVIEW

File: src/shogun/machine/BaggingMachine.cpp [hunk 1]
```diff
@@ -189,7 +189,7 @@ bool BaggingMachine::train_machine(const std::shared_ptr<Features>& data, const
 		}
 		
 
-#pragma omp critical
```
Comment (gf712) [inline src/shogun/machine/BaggingMachine.cpp]: you should rebase

File: src/shogun/classifier/svm/WDSVMOcas.cpp [hunk 2]
```diff
@@ -158,37 +156,29 @@ int32_t WDSVMOcas::set_wd_weights()
...
-	{
-		if (data->get_feature_class() != C_STRING ||
-				data->get_feature_type() != F_BYTE)
-		{
-			error("Features not of class string type byte");
-		}
-		set_features(std::static_pointer_cast<StringFeatures<uint8_t>>(data));
-	}
+	// require((data->get_feature_class() == C_STRING || data->get_feature_type() == F_BYTE)),
+	// 	 "Features not of class string type byte");
```
Comment (gf712) [inline src/shogun/classifier/svm/WDSVMOcas.cpp]: why commented out?

File: src/shogun/machine/Machine.cpp [hunk 3]
```diff
@@ -101,21 +89,6 @@ bool Machine::train(
...
-            error("Invalid label for {}", get_name());
-
-	m_labels = lab;
-    }
-}
-
-std::shared_ptr<Labels> Machine::get_labels()
-{
-	return m_labels;
-}
```
Comment (gf712) [inline src/shogun/machine/Machine.cpp:118]: \o/

File: src/shogun/machine/Machine.cpp [hunk 4]
```diff
@@ -33,26 +32,11 @@ Machine::~Machine()
...
 
 	if (support_feature_dispatching())
 	{
-		require(data != NULL, "Features not provided!");
-		require(
-		    data->get_num_vectors() == m_labels->get_num_labels(),
-		    "Number of training vectors ({}) does not match number of "
-		    "labels ({})",
-		    data->get_num_vectors(), m_labels->get_num_labels());
-
```
Comment (gf712) [inline src/shogun/machine/Machine.cpp:55]: these checks are not needed?

File: src/shogun/machine/NonParametricMachine.h [hunk 5]
```diff
@@ -39,17 +38,31 @@ namespace shogun
...
+         * @return labels
+         */
+        virtual std::shared_ptr<Labels> get_labels()
+		{
+			return m_labels;
+		}
 	protected:
 		std::shared_ptr<Features> m_features;
 
 		// TODO
```
Comment (gf712) [inline src/shogun/machine/NonParametricMachine.h]: can remove the TODO?

File: src/shogun/machine/Pipeline.h [hunk 6]
```diff
@@ -120,8 +120,34 @@ namespace shogun
...
+						? transformer->fit(current_data, args...)
+						: transformer->fit(current_data);
+					current_data = transformer->transform(current_data);
+					
+				}
+				else
+				{
+					auto machine = shogun::get<std::shared_ptr<Machine>>(stage.second);
+					machine->train(current_data, args...);
+
```
Comment (gf712) [inline src/shogun/machine/Pipeline.h]: remove extra space

File: src/shogun/machine/Pipeline.h [hunk 7]
```diff
@@ -120,8 +120,34 @@ namespace shogun
...
+			for (auto&& stage : m_stages)
+			{
+				if (holds_alternative<std::shared_ptr<Transformer>>(stage.second))
+				{
+					auto transformer = shogun::get<std::shared_ptr<Transformer>>(stage.second);
+					transformer->train_require_labels()
+						? transformer->fit(current_data, args...)
+						: transformer->fit(current_data);
+					current_data = transformer->transform(current_data);
+
```
Comment (gf712) [inline src/shogun/machine/Pipeline.h]: remove extra space

File: src/shogun/neuralnets/NeuralNetwork.cpp [hunk 8]
```diff
@@ -684,16 +690,9 @@ SGMatrix<float64_t> NeuralNetwork::labels_to_matrix(const std::shared_ptr<Labels
...
-	if (m_labels==NULL)
-		return PT_MULTICLASS;
-
-	if (m_labels->get_label_type() == LT_BINARY)
-		return PT_BINARY;
-	else if (m_labels->get_label_type() == LT_REGRESSION)
-		return PT_REGRESSION;
-	else return PT_MULTICLASS;
+	// // problem type depends on the type of labels given to the network
+	// // if no labels are given yet, just return PT_MULTICLASS
```
Comment (gf712) [inline src/shogun/neuralnets/NeuralNetwork.cpp]: I think you can remove these comments

File: tests/unit/multiclass/tree/RandomForest_unittest.cc [hunk 9]
```diff
@@ -147,6 +147,7 @@ TEST_F(RandomForestTest, classify_non_nominal_test)
 
 	std::shared_ptr<Evaluation> eval=std::make_shared<MulticlassAccuracy>();
 	c->put(RandomForest::kOobEvaluationMetric, eval);
+
```
Comment (gf712) [inline tests/unit/multiclass/tree/RandomForest_unittest.cc]: remove extra space

**Comment by gf712** [comment]
Cool! I guess once this is merged we can start the PR against the develop branch :)

File: tests/unit/machine/Pipeline_unittest.cc [hunk 10]
```diff
@@ -50,7 +50,8 @@ TEST_F(PipelineTest, fit_predict)
 	                    ->then(machine);
 
 	// no labels given
-	EXPECT_THROW(pipeline->train(features), ShogunException);
+	//EXPECT_THROW(pipeline->train(features), ShogunException);
+	//pipeline->train(features, labels);
```
Comment (gf712) [inline tests/unit/machine/Pipeline_unittest.cc:54]: what happens here?

Comment (LiuYuHui) [reply tests/unit/machine/Pipeline_unittest.cc:54]: now pipeline have two `train` method, original `train` will not throw a exception.

File: src/shogun/machine/Machine.cpp (hunk 4 above)
Comment (LiuYuHui) [reply src/shogun/machine/Machine.cpp:55]: move these checks to `NonParametricMachine::train`

**Comment by gf712** [comment]
@LiuYuHui Can I merge this?

**Comment by LiuYuHui** [comment]
> @LiuYuHui Can I merge this?

@gf712 yes :D

**Pull request merged**